HW2/P2/parallel_vector.pyx

```diff
@@ -107,9 +113,23 @@ cpdef move_data_medium_grained(np.int32_t[:] counts,
    ##########
    with nogil:
        for r in range(repeat):
-           for idx in range(src.shape[0]):
+           for idx in prange(src.shape[0], num_threads = 4):
                if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+                   Nlocks_src = (src[idx] + N - 1) / N
+                   Nlocks_dest = (dest[idx] + N - 1) / N
+                   if Nlocks_src == Nlocks_dest:
+                       acquire(&locks[Nlocks_dest])
+                       counts[dest[idx]] += 1
+                       counts[src[idx]] -= 1
+                       release(&locks[Nlocks_dest])
+                   else:
+                      acquire(&locks[Nlocks_dest])
+                      counts[dest[idx]] += 1
+                      release(&locks[Nlocks_dest])
+                      acquire(&locks[Nlocks_src])
+                      counts[src[idx]] -= 1
+                      release(&locks[Nlocks_src])   
+
```

Comment:
Looks good. I'm not sure if it is necessary to transform `src[idx]` and `dest[idx]` into the `Nlocks` version (mine works without it) but if it works that should be fine too.